Fix #8812 - Pass model type so search_for is called on Host

roidelapluie opened the pull request:
This PR is a rebase of GH-2013

At least on version 1.6.1, the absence of this second parameter leads to a
runtime crash when it's time to validate if the current user (non-admin) is
allowed to perform a power operation on given a host via the APIv2.

The root cause of the crash is basically that search_for is called on
Host::Base by app/services/authorizer.rb:50.

Not sure this is necessary in newer versions, up to the developers to figure
it out :)

lzap commented:
My initial idea was - holy moly, someone is programming Foreman with PL 2313 ;-)

![](http://iteadstudio.com/store/images/produce/Tool/MeasurementDiagnostic/USB2TLL/USBtoTLL1.jpg)

lzap commented:
FYI it is not required to create new PR when rebasing. Feel free to force push into your existing branch, github handles this fine.

roidelapluie commented:
@lzap I am not the author of the first PR, so I can not force push to someone else's repository. But I am really interested in getting that patch merged for 1.8/1.7..4

lzap commented:
> @lzap I am not the author of the first PR, so I can not force push to someone else's repository. But I am really interrested in getting that patch merged for 1.8/1.7..4

Oh sorry I missed that.

Thanks for the contributions. I expect folks who were doing initial
review will carry on with this one.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented:
[test]

roidelapluie commented:
@domcleal Can you retrigger the tests? Thanks

domcleal commented:
[test]

dLobatog commented:
@roidelapluie could you take at look at fixing the failing test? Thanks!

dLobatog commented:
[test] because of the latest enabled cops

roidelapluie commented:
Would be great to see this in 1.8.

domcleal commented:
[test]

Commit:
Fixes #8812 - Pass model type so search_for is called on Host

At least on version 1.6.1, the absence of this second parameter leads to a
runtime crash when it's time to validate if the current user (non-admin) is
allowed to perform a power operation on given a host via the APIv2.

The root cause of the crash is basically that search_for is called on
Host::Base by app/services/authorizer.rb:50.

roidelapluie commented:
@eLobato I have written a test but I can not get it to work locally

roidelapluie commented:
it does not work because in app/models/user.rb line 320    cached_roles.detect {|role| role.allowed_to?(action)}.present

cached_roles is empty. i have no idea why. when I connect to the test server I see that my fixtures are correct.

dLobatog commented:
I tried and I'm not sure why `cached_roles` is not being populated. However we're trying to get rid of using fixtures everywhere (especially in this case when it's a user only used once in the test base). 

I spent some time changing your code to use Factories and tests pass just fine :smile: , 

Here is [the patch](https://gist.github.com/eLobato/ef997b9030650fa610eb) !

I think this is probably ready to merge after applying it. Apply it, let's see what Mr. Jenkins thinks and if it's a :green_apple: let's :shipit: :smile: 

Thanks again @roidelapluie !

roidelapluie commented:
Merged your patch :)

roidelapluie commented:
@eLobato :green:

dLobatog commented:
Merged as e02a2ff23dfec7361c688b469461cc1e926bba23, thanks to @roidelapluie and special thanks to @nbarrientos for the original patch!
